File: p2p_library/src/main/java/com/tambapps/p2p/file_sharing/task/ReceivingTask.java
package com.tambapps.p2p.file_sharing.task;

import com.tambapps.p2p.file_sharing.Peer;
import com.tambapps.p2p.file_sharing.listener.ReceivingListener;
import com.tambapps.p2p.file_sharing.listener.TransferListener;

import java.io.DataInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.channels.SocketChannel;

/**
 * A P2P Receive task
 */
public class ReceivingTask extends SharingTask {

    private final FileProvider fileProvider;
    private volatile File outputFile = null;

    public ReceivingTask(File file) {
        this(null, file);
    }

    public ReceivingTask(FileProvider fileProvider) {
        this(null, fileProvider);
    }

    public ReceivingTask(TransferListener transferListener, File file) {
        this(transferListener, name -> file);
    }

    public ReceivingTask(TransferListener transferListener, FileProvider fileProvider) {
        super(transferListener);
        this.fileProvider = fileProvider;
    }

    public void receiveFrom(InetAddress address, int port) throws IOException {
        receiveFrom(Peer.of(address, port));
    }

    public void receiveFrom(Peer peer) throws IOException {
        try (SocketChannel socketChannel = SocketChannel.open()) {
            socketChannel.connect(new InetSocketAddress(peer.getIp(), peer.getPort()));
            while (!socketChannel.isConnected()) {
                try {
                    Thread.sleep(2000);
                } catch (InterruptedException e) {
                    cancel();
                    return;
                }
                if (isCanceled()) {
                    return;
                }
            }
            if (isCanceled()) {
                return;
            }
            try (Socket socket = socketChannel.socket();
                 DataInputStream dis = new DataInputStream(socket.getInputStream())) {
                long totalBytes = dis.readLong();
                int bufferSize = dis.readInt();
                String fileName = readName(dis);
                if (transferListener != null) {
                    transferListener.onConnected(peer, Peer.of(socket.getInetAddress(), socket.getPort()),
                      fileName, totalBytes);
                }

                outputFile = fileProvider.newFile(fileName);
                try (FileOutputStream fos = new FileOutputStream(outputFile)) {
                    share(bufferSize, dis, fos, totalBytes);
                }
                if (transferListener != null && transferListener instanceof ReceivingListener) {
                    ((ReceivingListener) transferListener).onEnd(outputFile);
                }
            }
        }

    }

    private String readName(DataInputStream dis) throws IOException {
        char[] name = new char[dis.readInt()];
        for (int i = 0; i < name.length; i++) {
            name[i] = dis.readChar();
        }
        return new String(name);
    }

    public File getOutputFile() {
        return outputFile;
    }

}

File: p2p_library/src/main/kotlin/com/tambapps/p2p/fandem/task/ReceivingTask.kt
package com.tambapps.p2p.fandem.task

import com.tambapps.p2p.fandem.Peer
import com.tambapps.p2p.fandem.io.CustomDataInputStream
import com.tambapps.p2p.fandem.listener.ReceivingListener
import com.tambapps.p2p.fandem.listener.TransferListener
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.channels.SocketChannel

/**
 * A P2P Receive task
 */
class ReceivingTask(transferListener: TransferListener?, private val fileProvider: FileProvider) : SharingTask(transferListener) {
  @Volatile
  var outputFile: File? = null
    private set

  constructor(file: File?) : this(null, file) {}
  constructor(fileProvider: FileProvider) : this(null, fileProvider) {}
  constructor(transferListener: TransferListener?, file: File?) : this(transferListener, FileProvider { file }) {
  }

  @Throws(IOException::class)
  fun receiveFrom(address: InetAddress, port: Int) {
    receiveFrom(Peer.of(address, port))
  }

  @Throws(IOException::class)
  fun receiveFrom(peer: Peer) {
    SocketChannel.open().use { socketChannel ->
      socketChannel.connect(InetSocketAddress(peer.ip, peer.port))
      while (!socketChannel.isConnected) {
        try {
          Thread.sleep(2000)
        } catch (e: InterruptedException) {
          cancel()
          return
        }
        if (isCanceled) {
          return
        }
      }
      if (isCanceled) {
        return
      }
      socketChannel.socket().use { socket ->
        CustomDataInputStream(socket.getInputStream()).use { dis ->
          val totalBytes = dis.readLong()
          val bufferSize = dis.readInt()
          val fileName = dis.readString()
          transferListener?.onConnected(peer, Peer.of(socket.inetAddress, socket.port),
              fileName, totalBytes)
          outputFile = fileProvider.newFile(fileName)
          FileOutputStream(outputFile).use { fos -> share(bufferSize, dis, fos, totalBytes) }
          if (transferListener != null && transferListener is ReceivingListener) {
            transferListener.onEnd(outputFile)
          }
        }
      }
    }
  }

}